Framework/Beamline/inc/MantidBeamline/ComponentInfo.h

```diff
@@ -36,22 +38,89 @@ namespace Beamline {
   Code Documentation is available at: <http://doxygen.mantidproject.org>
 */
 class MANTID_BEAMLINE_DLL ComponentInfo {
-
 private:
+  class Range {
+  private:
+    const std::vector<size_t>::const_iterator m_begin;
+    const std::vector<size_t>::const_iterator m_end;
+
+  public:
+    Range(std::vector<size_t>::const_iterator &&begin,
+          std::vector<size_t>::const_iterator &&end)
+        : m_begin(std::move(begin)), m_end(std::move(end)) {}
+    bool empty() const { return m_begin == m_end; }
+    auto begin() const -> decltype(m_begin) { return m_begin; }
+    auto end() const -> decltype(m_end) { return m_end; }
+  };
+
+  Range detectorRangeInSubtree(const size_t index) const;
+  Range componentRangeInSubtree(const size_t index) const;
+
   boost::shared_ptr<const std::vector<size_t>> m_assemblySortedDetectorIndices;
-  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>> m_ranges;
-  size_t m_size = 0;
+  /// Contains only indices of non-detector components
+  boost::shared_ptr<const std::vector<size_t>> m_assemblySortedComponentIndices;
+  /// Ranges of component ids that are contiguous blocks of detectors.
+  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+      m_detectorRanges;
+  /// Ranges of component ids that are contiguous blocks of components.
+  boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+      m_componentRanges;
+  boost::shared_ptr<const std::vector<size_t>> m_parentIndices;
+  Mantid::Kernel::cow_ptr<std::vector<Eigen::Vector3d>> m_positions;
+  Mantid::Kernel::cow_ptr<std::vector<Eigen::Quaterniond>> m_rotations;
+  const size_t m_size = 0;
+  const int64_t m_sourceIndex = -1;
+  const int64_t m_sampleIndex = -1;
+  DetectorInfo *m_detectorInfo; // ExperimentInfo is the owner.
```

Comment:
The wrappers `Geometry::DetectorInfo` in this case, are now the actual owners. I'll update comment.